### PR TITLE
Fix boost 1.73.0 support

### PR DESCRIPTION
**Pull request by TristanCacqueray**
## Purpose and Motivation

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1842012

It seems like code outside Boost should not include "detail" headers.
This change replaces the include by the correct endian header.

## Types of changes

- Bug fix

## To-do list

- [ ] Code is tested
- [ ] All tests are passing
- [ ] Updated documentation
- [ ] This PR is ready for review

### CODE REVIEW

**Comment by TristanCacqueray** [comment]
Not sure what is the strategy with regards to bundled external libraries, this change picked the new boost/endian/endian.hpp from https://github.com/boostorg/endian/blob/develop/include/boost/endian.hpp

**Comment by brianlheim** [comment]
hey @TristanCacqueray , i am not quite sure how this is fixing 1.73.0 support, can you please explain what you are trying to do and how this solves the problem? this would have been great information to include under the "Purpose and motivation" section of the pull request template.

**Comment by TristanCacqueray** [comment]
Oops it seems like `git-pull-request` altered the top comment when i pushed the second commit. I restored the initial proposal. The problem this solves is:

```
/builddir/build/BUILD/SuperCollider-Source/external_libraries/oscpack_build.cpp:1:10: fatal error: boost/detail/endian.hpp: No such file or directory
    1 | #include <boost/detail/endian.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by brianlheim** [comment]
no worries! are you building with system boost or vendored boost (i.e., did you use `-DSYSTEM_BOOST=ON`?)

**Comment by TristanCacqueray** [comment]
Yes this is happening when using SYSTEM_BOOST=ON with 1.7.30 installed. The spec file (describing how supercollider is built for Fedora) is: https://src.fedoraproject.org/rpms/supercollider/blob/master/f/supercollider.spec#_96

**Comment by TristanCacqueray** [comment]
Here are the reasons why it is built using system libraries: https://fedoraproject.org/wiki/Bundled_Libraries . Though if that is an issue, we could keep use the bundled libraries instead.

**Comment by brianlheim** [comment]
ok, thanks. no, SC should be able to build with the latest release of boost. in fact i've been meaning to update the version used in develop for awhile.

just curious, have you tried this build with SUPERNOVA=ON? that is what led to #4914

**Comment by TristanCacqueray** [comment]
IIUC this is already building with supernova without compilation issue. Here is the build logs: https://kojipkgs.fedoraproject.org//work/tasks/8018/45168018/build.log showing a single compilation failure because of the missing header.

Note that is happening with supercollider version 3.10.4

**Comment by brianlheim** [comment]
interesting! OK, thanks.

i looked around a bit and i think the correct macros are actually [`BOOST_ENDIAN_*_BYTE`](https://www.boost.org/doc/libs/1_73_0/libs/predef/doc/index.html#_boost_endian) in the predef library. as best i can tell these have never been defined in endian, you're just getting lucky because the endian library pulls in `predef/detail/endian_compat.h`. `BOOST_*_ENDIAN` is available but not publicly documented; in fact this appears to have been removed in the latest version of predef as it was already deprecated (https://github.com/boostorg/predef/commit/aa6e232bf170ad8b856aff9e7c70334f77441c7f). perhaps that hasn't made its way into the main boost distribution yet.

could you change the include to `<boost/predef.h>` and use the `BOOST_ENDIAN_*_BYTE` macros instead?

see also https://stackoverflow.com/questions/53936546/what-is-the-difference-between-boost-endian-big-byte-and-boost-endian-big-wor

**Comment by TristanCacqueray** [comment]
Oh that makes sense, thank you for the feedback. Please note that I haven't tested this PR.

File: external_libraries/oscpack_build.cpp [hunk 1]
```diff
@@ -1,8 +1,8 @@
-#include <boost/detail/endian.hpp>
+#include <boost/predef.h>
 
-#ifdef BOOST_LITTLE_ENDIAN
+#ifdef BOOST_ENDIAN_LITTLE_BYTE
 #define OSC_HOST_LITTLE_ENDIAN
-#elif defined(BOOST_BIG_ENDIAN)
+#elif defined(BOOST_ENDIAN_BIG_BYTE)
```
Comment (brianlheim) [inline external_libraries/oscpack_build.cpp:5]: [per documentation](https://www.boost.org/doc/libs/1_73_0/libs/predef/doc/index.html#_using_the_predefs), these are always defined, but nonzero, so you should use #if and not #ifdef

Comment (TristanCacqueray) [reply external_libraries/oscpack_build.cpp]: Done in e48ffe6

**Comment by TristanCacqueray** [comment]
Thank you! Note that this patch is not enough and we also need to fix the usage of ` BOOST_THROW_EXCEPTION` as mentioned in #4914

**Pull request merged**